Update tyk-helm-tyk-stack.md (#3441)
TykTechnologies/tyk-docs: tyk-docs/content/deployment-and-operations/tyk-self-managed/deployment-lifecycle/installations/kubernetes/tyk-helm-tyk-stack.md

{{< warning success >}}
**Warning**

The new Helm Charts are in beta stage. Breaking changes may be introduced before stable release.
{{< /warning >}}

To deploy Tyk Self Managed (for single data center) using the new helm chart, please use [tyk-stack](https://github.com/TykTechnologies/tyk-charts/tree/main/tyk-stack) chart.

## Tyk Self Managed (Single Data Center)

`tyk-stack` provides the default deployment of Tyk Self Managed on a cluster. It will deploy all required Tyk components with the settings provided in the values.yaml file.

It includes:
- Tyk Gateway, an open source Enterprise API Gateway (supporting REST, GraphQL, TCP and gRPC protocols).
- Tyk Pump, an analytics purger that moves the data generated by your Tyk nodes to any back-end. Furthermore, it has all the required modifications to easily connect to Tyk Cloud or Multi Data Center (MDCB) control plane.

> - Tyk Dashboard, a license based component that provides graphical management interface and analytics platform for Tyk.

Reviewer: ```suggestion
Tyk Dashboard, a license based component that provides a graphical management interface and analytics platform for Tyk.
```

## Tyk Self Managed (Single Data Center)

`tyk-stack` provides the default deployment of Tyk Self Managed on a cluster. It will deploy all required Tyk components with the settings provided in the values.yaml file.

It includes:
- Tyk Gateway, an open source Enterprise API Gateway (supporting REST, GraphQL, TCP and gRPC protocols).
- Tyk Pump, an analytics purger that moves the data generated by your Tyk nodes to any back-end. Furthermore, it has all the required modifications to easily connect to Tyk Cloud or Multi Data Center (MDCB) control plane.

> - Tyk Dashboard, a license based component that provides GUI management interface and analytics platform for Tyk.

Reviewer 2: Since GUI = graphical user interface , GUI management interface sounds redundant.

Author: updated

> - Tyk Enterprise Portal, a full-fledged CMS-like system for API providers to publish, monetise and drive the adoption of APIs.

Reviewer 2: Shouldn't we have Developer portal here?

Author: ```suggestion
- Tyk Developer Enterprise Portal, a full-fledged CMS-like system for API providers to publish, monetise and drive the adoption of APIs.
```

|Tyk Gateway |true  | n/a                    |
|Tyk Pump    |true | global.components.pump |
|Tyk Dashboard| true| global.components.dashboard | 

> |Tyk Enterprise Portal| false| global.components.devPortal | 

Author: ```suggestion
| Tyk Developer Enterprise Portal | false| global.components.devPortal | 
```

>NOTE: Please make sure you are installing Mongo/Postgres versions that are supported by Tyk. Please refer to Tyk docs to get list of [supported versions]({{< ref "tyk-dashboard/database-options" >}}).

### Protect Confidential Fields with Kubernetes Secrets

In the `values.yaml` file, some fields are considered confidential, such as `APISecret`, connection strings, etc.
Declaring values for such fields as plain text might not be desired for all use cases. Instead, for certain fields,
Kubernetes secrets can be referenced, and Kubernetes by itself configures values based on the referred secret.

This section describes how to use Kubernetes secrets to declare confidential fields.

#### Tyk Dashboard Admin

> If Tyk Dashboard bootstrapping is enabled, Tyk Dashboard admin will be bootstrapped according to `global.adminUser` field.

Reviewer: Wen bootstrapped does it mean that it will use the privileges/permissions associated with the Dashboard admin user?

```suggestion
If Tyk Dashboard bootstrapping is enabled, Tyk Dashboard admin will be bootstrapped according to the `global.adminUser` field.
```

Author: It means we'll create the admin user in Database with the given user details

Author: ```suggestion
If Tyk Dashboard bootstrapping is enabled, Tyk Dashboard admin user will be created according to the `global.adminUser` field.
```

> `global.adminUser.lastName`, `global.adminUser.email` and `global.adminUser.password`.

> [!WARNING]
> If `global.adminUser.useSecretName` is in use, please add all keys mentioned below to the secret.

##### Admin First Name

It can be configured via `global.adminUser.firstName` as a plain text or Kubernetes secret which includes `adminUserFirstName` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Last Name

It can be configured via `global.adminUser.lastName` as a plain text or Kubernetes secret which includes `adminUserLastName` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Email

It can be configured via `global.adminUser.email` as a plain text or Kubernetes secret which includes `adminUserEmail` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Password

It can be configured via `global.adminUser.password` as a plain text or Kubernetes secret which includes `adminUserPassword` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

#### APISecret

> [`APISecret`](https://tyk.io/docs/tyk-oss-gateway/configuration/#secret) field configures a header value used in every

Reviewer: ```suggestion
The [`APISecret`](https://tyk.io/docs/tyk-oss-gateway/configuration/#secret) field configures a header value used in every
interaction with Tyk Gateway API.
```

##### Admin First Name

It can be configured via `global.adminUser.firstName` as a plain text or Kubernetes secret which includes `adminUserFirstName` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Last Name

It can be configured via `global.adminUser.lastName` as a plain text or Kubernetes secret which includes `adminUserLastName` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Email

It can be configured via `global.adminUser.email` as a plain text or Kubernetes secret which includes `adminUserEmail` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

##### Admin Password

It can be configured via `global.adminUser.password` as a plain text or Kubernetes secret which includes `adminUserPassword` key
in it. Then, this secret must be referenced via `global.adminUser.useSecretName`.

#### APISecret

The [`APISecret`](https://tyk.io/docs/tyk-oss-gateway/configuration/#secret) field configures a header value used in every
interaction with Tyk Gateway API.

> interaction with Tyk Gateway API.

Reviewer 2: duplicated row

Author: oops thanks for spotting

It can be configured via `global.secrets.APISecret` as a plain text or Kubernetes secret which includes `APISecret` key
in it. Then, this secret must be referenced via `global.secrets.useSecretName`.

```yaml
global:
    secrets:
        APISecret: CHANGEME
        useSecretName: "mysecret" # where mysecret includes `APISecret` key with the desired value.
```

#### AdminSecret

[`AdminSecret](https://tyk.io/docs/tyk-dashboard/configuration/#admin_secret) sets a secret for Admin API.

It can be configured via `global.secrets.AdminSecret` as a plain text or Kubernetes secret which includes `AdminSecret` 
key in it. Then, this secret must be referenced via `global.secrets.useSecretName`.

```yaml
global:
    secrets:
        useSecretName: "mysecret" # where mysecret includes `useSecretName` key with the desired value.
```

> [!NOTE]
> Once `global.secrets.useSecretName` is declared, it takes precedence over `global.secrets.APISecret` and `global.secrets.AdminSecret`.

#### Dashboard License

In order to refer Tyk Dashboard license through Kubernetes secret, please use `global.secrets.useSecretName`, where 
the secret should contain a key called `DashLicense`.

> #### Enterprise Portal License

Reviewer 2: Shouldn't we have Developer portal here?

Author: ```suggestion
#### Tyk Developer Enterprise Portal License
```

> In order to refer Tyk Enterprise Portal license through Kubernetes secret, please use `global.secrets.useSecretName`, 

Author: ```suggestion
In order to refer Tyk Developer Enterprise Portal license through Kubernetes secret, please use `global.secrets.useSecretName`, 
```

where the secret should contain a key called `EnterprisePortalLicense`.

> #### Enterprise Portal Admin Password

Reviewer 2: Shouldn't we have Developer portal here?

Author: ```suggestion
#### Tyk Developer Enterprise Portal Admin Password
```

> In order to refer Tyk Enterprise Portal's admin password through Kubernetes secret, 

Author: ```suggestion
In order to refer Tyk Developer Enterprise Portal's admin password through Kubernetes secret, 
```

please use `global.adminUser.useSecretName`, where the secret should contain a key called `adminUserPassword`.

> #### Enterprise Portal Storage Connection String

Reviewer 2: Shouldn't we have Developer portal here?

3. Set `gateway.tls.useDefaultTykCertificate` to false.
4. Set `gateway.tls.secretName` to the name of the newly created secret.

#### Enabling gateway autoscaling

> You can enable autoscaling of gateway by `--set tyk-gateway.gateway.autoscaling.enabled=true`. By default, it will enable `Horizontal Pod Autoscaler` resource with target average CPU utilisation at 60%, scaling between 1 and 3 instances. To customize those values you can modify below section of `values.yaml`:

Reviewer 2: "of the gateway" is correct

Author: updated

        password: "123456"
      # Dashboard Organisation information.
```

### Tyk Enterprise Portal Configurations

To enable Enterprise Portal, set `global.components.devPortal` to true, and configure below inside `tyk-dev-portal` section.

#### Enterprise Portal License (Required)

Enterprise Portal License is required. It can be set up in `tyk-dev-portal.license` or through secret `global.secrets.useSecretName`. The secret should contain a key called DeveloperPortalLicense.

```yaml
tyk-dev-portal:
  # Tyk Developer Portal license.
  license: ""
```

#### Storage Settings

> Enterprise Portal supports different storage options for the portal's CMS assets. Please see the [Enterprise Portal Storage settings] page for all the available options. Most of them can be setup in corresponding `storage` section:

Reviewer: Add reference link to enterprise portal storage settings page

Author: ```suggestion
Enterprise Portal supports different storage options for the portal's CMS assets. Please see the [Enterprise Portal Storage settings]({{<ref "/tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration/#portal-settings">}}) page for all the available options. Most of them can be setup in corresponding `storage` section:
```

    # Supported types: fs, s3, db
    type: "fs"
    database:
      dialect: "sqlite3"
      connectionString: "db/portal.db"
      enableLogs: false
      maxRetries: 3
      retryDelay: 5000
    # Configuration values for using s3 as storage for Tyk Developer Portal
    # In case you want to provide the key ID and access key via secrets please
    # refer to the existing secret inside the helmchart or the
    # .Values.global.secrets.useSecretName variable and a secret containing
    # the keys DeveloperPortalAwsAccessKeyId and respectively,
    # DeveloperPortalAwsSecretAccessKey
    s3:
      awsAccessKeyid: your-access-key
      awsSecretAccessKey: your-secret-key
      region: sa-east-1
      endpoint: your-portal-bucket
      bucket: https://s3.sa-east-1.amazonaws.com
      acl: private
      presign_urls: true
    persistence:
      # User can mount existing PVC to the Tyk Developer Portal
      # Make sure to change the kind to Deployment if you are mounting existing PVC 
      mountExistingPVC: ""
      storageClass: ""
      accessModes:
        - ReadWriteOnce
      size: 8Gi
      annotations: {}
      labels: {}
      selector: {}
```

#### Other Configurations

> For other [Enterprise Portal configurations]({{<ref "/tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration">}}), it can be set by using environment variables with `extraEnvs` fields, e.g.:

Reviewer: ```suggestions

Other [Enterprise Portal configurations]({{<ref "/tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration">}}) can be set by using environment variables with `extraEnvs` fields, e.g.:
```

Author: ```suggestion
Other [Enterprise Portal configurations]({{<ref "/tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration">}}) can be set by using environment variables with `extraEnvs` fields, e.g.:
```

        password: "123456"
      # Dashboard Organisation information.
```

> ### Tyk Enterprise Portal Configurations

Reviewer 2: Shouldn't we have Developer portal here?

Author: ```suggestion
### Tyk Developer Enterprise Portal Configurations
```

> To enable Enterprise Portal, set `global.components.devPortal` to true, and configure below inside `tyk-dev-portal` section.

Author: ```suggestion
To enable Tyk Developer Enterprise Portal, set `global.components.devPortal` to true, and configure below inside `tyk-dev-portal` section.
```

> #### Enterprise Portal License (Required)

Author: ```suggestion
#### Tyk Developer Enterprise Portal License (Required)
```

> Enterprise Portal License is required. It can be set up in `tyk-dev-portal.license` or through secret `global.secrets.useSecretName`. The secret should contain a key called DeveloperPortalLicense.

Author: ```suggestion
Tyk Developer Enterprise Portal License is required. It can be set up in `tyk-dev-portal.license` or through secret `global.secrets.useSecretName`. The secret should contain a key called DeveloperPortalLicense.
```

```yaml
tyk-dev-portal:
  # Tyk Developer Portal license.
  license: ""
```

#### Storage Settings

> Enterprise Portal supports different storage options for the portal's CMS assets. Please see the [Enterprise Portal Storage settings]({{<ref "tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration#portal-settings">}}) page for all the available options. Most of them can be setup in corresponding `storage` section:

Author: ```suggestion
Tyk Enterprise Portal supports different storage options for storing the portal's CMS assets such as images, theme files and Open API Specification files. Please see the [Enterprise Portal Storage settings]({{<ref "tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration#portal-settings">}}) page for all the available options. Helm chart supports the setting of following fields in values.yaml `storage` section:
```

```yaml
tyk-dev-portal:
  storage:
    # Configuration values for using an SQL database as storage for Tyk Developer Portal
    # In case you want to provide the connection string via secrets please
    # refer to the existing secret inside the helmchart or the
    # .Values.global.secrets.useSecretName variable
    # User can set the storage type for portal.

>     # Supported types: fs, s3, db

Reviewer 3: All types of storage are mutually exclusive. In other words, only one of them can be used at time. The below explanation makes it confusing to a user: it looks like I have to setup all of them. Moreover, if a user will copy that configuration, it guaranteed that the portal won't start.

Therefore:
- Please split this into three sections explaining each option separately: `fs`, `db` and `persistence`
- Also, please provide explanation about what each of the options means to the end-user:
 - `fs`: the portal's assets will be stored on the filesystem, therefore a persistent volume claim will be created.
 - `s3`: the portal's assets will be stored in a S3 volume.
 - `db`: the portal's assets will be stored in the main portal's database.

Author: updated

    type: "fs"

>     database:

Reviewer 3: Where is the `db` storage type?

Author: database. it corresponds to portal storage config:

PORTAL_STORAGE

Config file: Storage
Type: string
Options:

fs to use file system storage type;
db to use the portal’s main database. If the db is selected as a storage type, the portal application will create appropriate structure in the database that
s3 to use S3 volume for storing the portal assets.
Description: Defines which type of storage to use for the portal’s CMS assets. Not required. If it is not specified, the default value is fs.

      dialect: "sqlite3"
      connectionString: "db/portal.db"
      enableLogs: false
      maxRetries: 3

>       retryDelay: 5000

Reviewer 3: The database connection settings are not related to the storage configuration. A user has to specify connection setting to their database regardless which storage type is selected. Why are they in this section?

Author: @andrei-tyk

Reviewer 2: This is an oversight by me which was probably ported from the old version of the helmcharts. Thank you for the explanation @slega. We should create a ticket to update this @caroltyk

Author: this will be fixed in next release

Tyk Enterprise Developer Portal License is required. It can be set up in `tyk-dev-portal.license` or through secret `global.secrets.useSecretName`. The secret should contain a key called DevPortalLicense.

```yaml
tyk-dev-portal:
  # Tyk Developer Portal license.
  license: ""
```

#### Tyk Enterprise Developer Portal Database

By default, Tyk Enterprise Developer Portal use `sqlite3` to store portal metadata. If you want to use other SQL Database, please modify the section below.

```yaml
tyk-dev-portal:
  storage:
    database:
      dialect: “sqlite3”
      connectionString: “db/portal.db”
      enableLogs: false
      maxRetries: 3
      retryDelay: 5000
```

Note: Database settings will be moved out of the storage section in next release.

#### Storage Settings

Tyk Enterprise Developer Portal supports different storage options for storing the portal's CMS assets such as images, theme files and Open API Specification files. Please see the [Enterprise Developer Portal Storage settings]({{<ref "tyk-stack/tyk-developer-portal/enterprise-developer-portal/install-tyk-enterprise-portal/configuration#portal-settings">}}) page for all the available options. 

> If you use filesystem as storage, please set `tyk-dev-portal.storage.type` to `fs`, and configure `tyk-dev-portal.storage.persistence` to mount an existing persistent volume to Tyk Enterprise Developer Portal.

Reviewer: ```suggestion
If you use the file system as storage, please set `tyk-dev-portal.storage.type` to `fs`, and configure `tyk-dev-portal.storage.persistence` to mount an existing persistent volume to Tyk Enterprise Developer Portal.
```
